TypeScript_ESLint.snb.md

```diff
@@ -0,0 +1,131 @@
+
+In this post, we provide a hands-on, technical introduction to how this essential static analysis tool works by tracing through its key code paths. We'll cover the following:
+
+* How linting rules are defined
+* The use of dueling ASTs
+* A neat co-recursive switch pattern for AST traversal
+
+Read on!
+
+## Overview
+
+TypeScript ESLint is implemented as a collection of packages that build on ESLint. It has a few main modules, the most important of which are:
+
+* `eslint-plugin`, which defines the ESLint plugin and the linting rules.
+* `parser` and `typescript-estree`, which are responsible for parsing TypeScript source code into an ESLint-compatible form, mapping from TypeScript ASTs to ESLint's AST representation (ESTree) and providing an API to query the AST from the rules.
+
+## How rules work
+
+The purpose of the linter is to detect a set of patterns in source code that match common anti-patterns. These patterns are described in **rules**. The docs contain a [complete list of these rules](https://typescript-eslint.io/rules/).
+
+To dive into the code, let's start with the definition of a simple rule. The `no-for-in-array` rule disallows iterating over an array with a for-in loop. Here is an example of this pattern:
+
+https://sourcegraph.com/github.com/cypress-io/cypress/-/blob/packages/driver/src/cypress/proxy-logging.ts?L13
+
+This anti-pattern is quite widespread, because it feels intuitive to use the for-in syntax to traverse an array. Doing so, however, may visit the array elements out-of-order. Instead `array.forEach` is recommended.
+
+Let's dive into the definition of this rule in source. It is short enough to include here in its entirety:
+
+https://sourcegraph.com/github.com/typescript-eslint/typescript-eslint/-/blob/packages/eslint-plugin/src/rules/no-for-in-array.ts
+
+The rule definition contains metadata at the top and the actual logic of the rule below. Let's zoom in on that logic:
+
+https://sourcegraph.com/github.com/typescript-eslint/typescript-eslint/-/blob/packages/eslint-plugin/src/rules/no-for-in-array.ts?L22-41
+
+There may be some unfamiliar syntax for some here. This code might look like it's defining a function called `ForInStatement` within an object declaration, but it's really just a terse variant of declaring a function value for an object key, like this:
+
+```typescript
+{
+    ForInStatement: (node): void => { ... }
+}
+```
+
+This key will be used to match this rule with the name of an AST node type in the ESLint AST, so that the node argument will always have the type that the rule targets:
+
+https://sourcegraph.com/npm/typescript-eslint/types@540cca215af38ab1a49b6cde8bcb492f4b6e53c5/-/blob/dist/generated/ast-spec.d.ts?L75
+
+## Type-checking
+
+One distinction between TypeScript ESLint and vanilla ESLint is that we have access to the TypeScript type checker and can therefore define rules that make use of type information.
```

Comment:
Nit: it's a bit understated here, but the difference in power between type checked rules and vanilla rules can be _massive!_ Also splitting up the sentence, if you don't mind:

```suggestion
One powerful distinction between TypeScript ESLint and vanilla ESLint is that we have access to the TypeScript type checker.
Rules that make use of type checking are able to make informed decisions on nodes based on their type information.
```